packages/crypto/tests/rootberg_secp256k1.rs

```diff
@@ -0,0 +1,164 @@
+
+    pub fn sha256(data: &[u8]) -> [u8; 32] {
+        Sha256::digest(data).into()
+    }
+
+    pub fn keccak_256(data: &[u8]) -> [u8; 32] {
+        Keccak256::digest(data).into()
+    }
+}
+
+#[test]
+fn rootberg_ecdsa_secp256k1_sha256() {
+    let mut tested = 0;
+    let File { num_tests, tests } = read_file(SECP256K1_SHA256);
+    assert!(num_tests >= 423, "Got unexpected number of tests");
+
+    for test in tests {
+        assert_eq!(test.pubkey.len(), 65);
+
+        tested += 1;
+
+        assert_eq!(test.tc_id as usize, tested);
+        eprintln!("Test case ID: {}", test.tc_id);
+        let message_hash = hashers::sha256(&test.msg);
+
+        let signature = combine_signature(&test.sig);
+        match secp256k1_verify(&message_hash, &signature, &test.pubkey) {
+            Ok(valid) => assert_eq!(
+                test.valid || test.comment.contains("(s not in range 1 .. n//2)"),
+                valid
+            ),
+            Err(e) => {
+                assert!(!test.valid, "expected valid signature, got {:?}", e);
+            }
+        }
+
+        if test.valid {
+            let recovered =
+                secp256k1_recover_pubkey(&message_hash, &signature, test.sig.id).unwrap();
+            assert_eq!(recovered, test.pubkey);
+        }
+    }
+    assert_eq!(tested, num_tests);
+}
+
+#[test]
+fn rootberg_ecdsa_secp256k1_keccak256() {
+    let mut tested = 0;
+    let File { num_tests, tests } = read_file(SECP256K1_KECCAK256);
+    assert!(num_tests >= 263, "Got unexpected number of tests");
+
+    for test in tests {
+        assert_eq!(test.pubkey.len(), 65);
+
+        tested += 1;
+
+        assert_eq!(test.tc_id as usize, tested);
+        eprintln!("Test case ID: {}", test.tc_id);
+        let message_hash = hashers::keccak_256(&test.msg);
+
+        let signature = combine_signature(&test.sig);
+        match secp256k1_verify(&message_hash, &signature, &test.pubkey) {
+            Ok(valid) => assert_eq!(
+                test.valid || test.comment.contains("(s not in range 1 .. n//2)"),
+                valid
+            ),
+            Err(e) => {
+                assert!(!test.valid, "expected valid signature, got {:?}", e);
+            }
+        }
+
+        if test.valid {
+            let recovered =
+                secp256k1_recover_pubkey(&message_hash, &signature, test.sig.id).unwrap();
+            assert_eq!(recovered, test.pubkey);
+        }
+    }
+    assert_eq!(tested, num_tests);
+}
+
+fn combine_signature(sig: &Sig) -> Vec<u8> {
+    // the test data contains values with leading zeroes, which we need to ignore
+    let first_non_zero = sig
+        .r
+        .iter()
+        .enumerate()
+        .find(|(_, v)| **v != 0) // find first non-zero byte
+        .map(|(i, _)| i)
+        .unwrap_or_default(); // default to 0 if all zero
+    let r = &sig.r[first_non_zero..];
+    let first_non_zero = sig
+        .s
+        .iter()
+        .enumerate()
+        .find(|(_, v)| **v != 0)
```

Comment:
This is just a reimplementation of [`.position()`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.position)

```suggestion
        .position(|item| *v != 0)
```

Reply:
Oh, that's much nicer!